Refactor with OOP

Jake0Miller opened the pull request:
## Description
- Refactored code using a FileWriter class
- Minor tweaks, mostly cosmetic
- Added a PR template which should look familiar

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Refactor (non-breaking change which maintains existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Notes
Unable to test locally -- error present before and after these changes when attempting `gem install ez_gen` (possibly user error on my part):

```
ERROR:  While executing gem ... (NoMethodError) 
undefined method `invoke_with_build_args' for nil:NilClass
```

tylerpporter commented:
Looks like the issue is on your end. [This thread](https://stackoverflow.com/questions/23350388/cant-install-gems-because-undefined-method-invoke-with-build-args-for-nilni) might help you out.

tylerpporter commented on lib/ez_gen/file_writer.rb:
```diff
@@ -0,0 +1,165 @@
+
+    desc "Run single test file based on class name, ex: rake test_class person"
+    task :test_class do
+      @args.each { |arg| task arg.to_sym do ; end }
```
This is creating static Rake tasks that will only ever create or call the initial classes/files that the user creates. This Rakefile is meant to take `ARGV` from the user once they have generated their project so that the user can run whichever test they want and generate new tests and classes.

Jake0Miller replied:
I mixed up ARGV for this script and the ARGV it creates for the user of the files it writes.

Jake0Miller replied:
Ah, that's even more useful than I first thought! I mixed up the ARGV for ez_gen with the ARGV embedded in this rake file for later user input.

tylerpporter commented on lib/ez_gen/file_writer.rb:
```diff
@@ -0,0 +1,165 @@
+    desc "Generate new class and test, ex: rake g person"
+    task :g do
+      @args.each { |arg| task arg.to_sym do ; end }
+      ruby "./.scripts/generate.rb #{@lower_name}"
```
`@lower_name` will be `nil` and this Rake task will error

Jake0Miller replied:
This is the same mistake I made above--I was interpolating the args passed to ez_gen into this rake file, rather than allowing different args passed into the rake file after ez_gen creates the rake file.

tylerpporter commented on lib/ez_gen/file_writer.rb:
```diff
@@ -0,0 +1,165 @@
+    class_file.close
+  end
+
+  def write_test_file
+    test_file = File.open("test/#{@lower_name}_test.rb", 'w')
+    test_file.write(
+    "require './test/test_helper.rb'
+    require './lib/#{@lower_name}.rb'
+
+    class #{@upper_name}Test < Minitest::Test
+      def setup
+        @#{@lower_name} = #{@upper_name}.new
+      end
+
+      def test_it_exists
+        assert_instance_of #{@upper_name}, @#{@lower_name}
+      end
+    end"
+    )
+    test_file.close
+    test_file
+  end
```
This is formatting the generated test file incorrectly: 
![Screen Shot 2020-08-23 at 7 21 02 AM](https://user-images.githubusercontent.com/58486125/90979458-654df000-e512-11ea-91bf-b4128fd20134.png)

It's not major (it will still execute for the user) but it's not ideal. Let's figure out a way to fix this indentation issue. 
 (This is happening with all files generated from within the `FileWriter` class)

Jake0Miller replied:
Unfortunate side effect of multi-line strings inside a method... If I can get the gem working I'll be able to test it, but likely solutions are either removing the indentation in `FileWriter` multi-line strings or importing all multi-line strings from another file without indentation.

I'll try the first fix and see how you like the look of it.

Jake0Miller replied:
Is this a satisfactory solution to the indentation in my most recent commit?

tylerpporter commented on lib/ez_gen/generator.rb:
```diff
@@ -1,179 +1,39 @@
 require 'fileutils'
 require 'pastel'
+require 'file_writer'
```
This needs either needs to be a `require_relative` or `require` the full project path

Jake0Miller replied:
Ah, thanks--I haven't worked in Ruby in a bit! Need to keep these skills sharp

Jake0Miller commented:
I had to install OpenSSL, then ran into an error that SimpleCov requires Ruby >= 2.5.0. My machine defaults to 2.4.1, which is what we were all supposed to use when I was at Turing, but I switched to a newer version with rbenv and then the gem installed with `rake install` as suggested. I had to `gem install mocha` as well after switching Ruby versions.

Do you want to add these requirements to a list of dependencies?

From there I was able to test and confirm the formatting (indentation) is now fixed.

tylerpporter commented:
Great! Yes, good call, we should definitely add those requirements to the list of dependencies.
